### PR TITLE
#184651082 Handling payments with stripe

### PR DESCRIPTION
**Pull request by princoo**
#### What does this PR do?
Enable the payments on a product to be made

#### Description of Task to be completed?
- ensure the buyer makes payment for a particular product
- ensures the transaction(order) status is stored in the database

#### How should this be manually tested?
- as a buyer you need to be logged in
- to make a transaction you need to make a `post` request on endpoint `/checkout/payment/{orderId}`
- once the required validations a meet and the payment details are valid, the transaction will happen
- if a payment requires additional authorization you will be redirected to the page for confirmation
- once you authorize the payment it will be successful
- if you cancel the authorization the payment will fail and you will need to provide payment details
-if  some request body data are not provide the payment will not go through

#### What are the relevant pivotal tracker/Trello stories?
#184651082

### CODE REVIEW

**Review by Edcreation — APPROVED** [review]

File: src/database/migrations/20230401104008-create-products.js
Comment (AdrineUWERA) [inline src/database/migrations/20230401104008-create-products.js:1]: U didn't have to edit this migration. Eddy created a migration (modify_collectionId) to fix this.

File: src/middleware/order/isOrderExists.js [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+const OrderExists = async (req, res, next) => {
+  const { orderId } = req.params;
+  const order = await orderService.getOrderById(orderId);
+  if (order) {
+    req.orderAmount = order.totalPrice;
+    return next();
+  }
+  return res
+    .status(400)
+    .json({ code: '400', Message: 'Failed', errro: 'Order does not exist' });
```
Comment (AdrineUWERA) [inline src/middleware/order/isOrderExists.js:12]: correct the type and case

**Review by AdrineUWERA — CHANGES_REQUESTED** [review]

**Review by mirindi19 — APPROVED** [review]

File: src/controllers/payment.controller.js [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+  paymentMethod,
+  charge,
+  stripeListener,
+  sendEmails,
+} from '../utils';
+import { orderService } from '../services';
+import { stripe } from '../helpers';
+
+const { ADMIN_EMAIL } = process.env;
+const makePayment = async (req, res) => {
```
Comment (ishimwe5555) [inline src/controllers/payment.controller.js:12]: in this function, if an error occurs, it is not caught, and the response is not handled appropriately. Instead of throwing an error, it is better to return a meaningful error message to the client.

File: src/controllers/payment.controller.js [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+        const status = 'incomplete';
+        const update = await orderService.updateOrderStatus(
+          status,
+          data.metadata.orderId
+        );
+        return update;
+      }
+      if (data.status === 'requires_payment_method') {
+        const status = 'failed';
+        const update = await orderService.updateOrderStatus(
```
Comment (ishimwe5555) [inline src/controllers/payment.controller.js:69]: The updateOrderStatus function is called multiple times in the webhook function with different status values. It would be better to create a separate function for updating the order status and reuse it. As in separate functions to update the to different statuses i.e: succeeded, incomplete and failed so that they can be independently reusable too.

**Review by ishimwe5555 — CHANGES_REQUESTED** [review]

**Review by AdrineUWERA — APPROVED** [review]

**Review by ishimwe5555 — APPROVED** [review]

**Review by AdrineUWERA — APPROVED** [review]

**Review by munezerobagira — APPROVED** [review]

**Review by JosephNgabo — APPROVED** [review]

File: src/public/index.html [hunk 4]
```diff
@@ -1,43 +1,42 @@
...
+          query: { authToken: token },
+        });
+        socket.on('unauthorized', (alerting) => {
+          alert(`Login First ${alerting}`);
+        });
+        socket.emit('join');
+        socket.id = user.data.id;
+        const userId = socket.id;
+        socket.on('joined', (data) => {
+          console.log(data);
```
Comment (munezerobagira) [inline src/public/index.html:34]: why did you change this

File: src/controllers/product.controller.js [hunk 5]
```diff
@@ -52,6 +52,7 @@ const addproduct = async (req, res) => {
     req.body;
   const img = req.files;
   let url = [];
+  // try {
```
Comment (munezerobagira) [inline src/controllers/product.controller.js:55]: why did you  comment this

File: src/database/migrations/20230401104008-create-products.js [hunk 6]
```diff
@@ -40,8 +40,8 @@ module.exports = {
         references: {
           model: 'collections',
           key: 'id',
-          onDelete: 'CASCADE',
         },
+        onDelete: 'CASCADE',
```
Comment (munezerobagira) [inline src/database/migrations/20230401104008-create-products.js:44]: don't edit migration

**Review by munezerobagira — CHANGES_REQUESTED** [review]

**Review by munezerobagira — APPROVED** [review]
